src/library_exceptions.js

```diff
@@ -102,33 +206,28 @@ var LibraryExceptions = {
 #endif
   },
 
-  __cxa_increment_exception_refcount__deps: ['__exception_addRef', '__exception_deAdjust'],
+  __cxa_increment_exception_refcount__deps: ['__exception_addRef', 'ExceptionInfo'],
   __cxa_increment_exception_refcount: function(ptr) {
-    ___exception_addRef(___exception_deAdjust(ptr));
+    if (!ptr) return;
+    ___exception_addRef(new _ExceptionInfo(ptr));
   },
 
-  __cxa_decrement_exception_refcount__deps: ['__exception_decRef', '__exception_deAdjust'],
+  __cxa_decrement_exception_refcount__deps: ['__exception_decRef', 'ExceptionInfo'],
   __cxa_decrement_exception_refcount: function(ptr) {
-    ___exception_decRef(___exception_deAdjust(ptr));
+    if (!ptr) return;
+    ___exception_decRef(new _ExceptionInfo(ptr));
   },
 
   // Here, we throw an exception after recording a couple of values that we need to remember
   // We also remember that it was the last exception thrown as we need to know that later.
   __cxa_throw__sig: 'viii',
-  __cxa_throw__deps: ['__exception_infos', '__exception_last', '_ZSt18uncaught_exceptionv'],
+  __cxa_throw__deps: ['ExceptionInfo', '__exception_last', '_ZSt18uncaught_exceptionv'],
   __cxa_throw: function(ptr, type, destructor) {
 #if EXCEPTION_DEBUG
     err('Compiled code throwing an exception, ' + [ptr,type,destructor]);
 #endif
-    ___exception_infos[ptr] = {
-      ptr: ptr,
-      adjusted: [ptr],
-      type: type,
-      destructor: destructor,
-      refcount: 0,
-      caught: false,
-      rethrown: false
-    };
+    var info = new _ExceptionInfo(ptr);
+    info.init(type, destructor);
```

Comment:
It's not clear to me why `init` is called here but not in other places. Can we add the two parameters to the constructor? Or if there is a fundamental reason, perhaps `init` needs a more meaningful name or comment.

Reply:
Here it is first created and initialized, all the other places just wrap the pointer to previously created ExceptionInfo instance. Will add the parameters to the constructor.

Reply:
Or may be leave it as is with some comment? Otherwise it will look like `new _ExceptionInfo(ptr, type, destructor);` thus relying purely on constructor side effects which does not look nice.

Reply:
Hmm, interesting.

How about the constructor has no side effects, and after it we can either `create` to create a new underlying instance, or `wrap` to wrap an existing one? That seems like it would be clear.